src/log.c

```diff
@@ -36,70 +36,378 @@
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
+#include <inttypes.h>
 
+#include <rdma/fi_errno.h>
 #include "fi.h"
 #include "fi_log.h"
 
+#define ARRAY_SIZE(x) (sizeof((x)) / sizeof((x)[0]))
+
+static struct slist provider_handles;
+static struct log_providers provider_list;
+
+const char *subsystems[] = {
+	[FABRIC] = "fabric",
+	[DOMAIN] = "domain",
+	[EP_CM] = "ep_cm",
+	[EP_DM] = "ep_dm",
```

Comment:
Maybe use EP_CTRL and EP_DATA?  Or just CM and EP?  The use of EP_DM isn't intuitive.